digitalocean/resource_digitalocean_database_firewall.go

```diff
@@ -0,0 +1,186 @@
+package digitalocean
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/digitalocean/godo"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+)
+
+func resourceDigitalOceanDatabaseFirewall() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDigitalOceanDatabaseFirewallCreate,
+		Read:   resourceDigitalOceanDatabaseFirewallRead,
+		Update: resourceDigitalOceanDatabaseFirewallUpdate,
+		Delete: resourceDigitalOceanDatabaseFirewallDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceDigitalOceanDatabaseFirewallImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"cluster_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.NoZeroValues,
+			},
+
+			"rule": {
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"type": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								"ip_addr",
+								"droplet",
+								"k8s",
+								"tag",
+							}, false),
+						},
+
+						"value": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.NoZeroValues,
+						},
+
+						"uuid": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+
+						"created_at": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceDigitalOceanDatabaseFirewallCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*CombinedConfig).godoClient()
+	clusterID := d.Get("cluster_id").(string)
+
+	rules := buildDatabaseFirewallRequest(d.Get("rule").(*schema.Set).List())
+
+	_, err := client.Databases.UpdateFirewallRules(context.TODO(), clusterID, &rules)
+	if err != nil {
+		return fmt.Errorf("Error creating DatabaseFirewall: %s", err)
+	}
+
+	d.SetId(resource.PrefixedUniqueId(clusterID + "-"))
+
+	return resourceDigitalOceanDatabaseFirewallRead(d, meta)
+}
+
+func resourceDigitalOceanDatabaseFirewallRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*CombinedConfig).godoClient()
+	clusterID := d.Get("cluster_id").(string)
+
+	rules, _, err := client.Databases.GetFirewallRules(context.TODO(), clusterID)
+	if err != nil {
+		return fmt.Errorf("Error retrieving DatabaseFirewall: %s", err)
+	}
+
+	if err := d.Set("rule", flattenDatabaseFirewallRules(rules)); err != nil {
+		return fmt.Errorf("[DEBUG] Error setting DatabaseFirewall rule error: %#v", err)
```

Comment:
Could we maybe just return directly from this set (Since it is the line with the last error) instead of a custom error message? Wouldn't the set failure be clear enough?